### PR TITLE
Begin adding recursive CTE to tree queries

### PR DESCRIPTION
**Pull request by realVinayak**
In tree viewer, tree stats refer to the count of collection object associated with tree nodes sharing a parent ParentID. Backend currently does this by making a flat join "depth" number of times. So, starting from the parent, it will make "depth" joins. A problem with this approach is since children are many-to-one to parent, the rows in all the tables (except the last one) could get duplicated - hence, we need to have distinct determination / collectionobject for getting counts. This leads to a major performance problem. I've tested these queries, and they sometimes don't even finish in 30 minutes if tree gets too large or co/det table gets too big. Using a recursive-cte approach fixes this by making a join only when it is needed and queries finish in less than a second. 

For example, a query that looks like this using flat join : 

```
SELECT 
  taxon_1.`TaxonID` AS `taxon_1_TaxonID`, 
  count(
    DISTINCT determination_1.`DeterminationID`
  ) AS count_1, 
  count(
    DISTINCT determination_1.`DeterminationID`
  ) + count(
    DISTINCT determination_2.`DeterminationID`
  ) + count(
    DISTINCT determination_3.`DeterminationID`
  ) + count(
    DISTINCT determination_4.`DeterminationID`
  ) + count(
    DISTINCT determination_5.`DeterminationID`
  ) + count(
    DISTINCT determination_6.`DeterminationID`
  ) + count(
    DISTINCT determination_7.`DeterminationID`
  ) + count(
    DISTINCT determination_8.`DeterminationID`
  ) + count(
    DISTINCT determination_9.`DeterminationID`
  ) + count(
    DISTINCT determination_10.`DeterminationID`
  ) AS anon_1 
FROM 
  taxon AS taxon_1 
  LEFT OUTER JOIN taxon AS taxon_2 ON taxon_2.`ParentID` = taxon_1.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_3 ON taxon_3.`ParentID` = taxon_2.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_4 ON taxon_4.`ParentID` = taxon_3.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_5 ON taxon_5.`ParentID` = taxon_4.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_6 ON taxon_6.`ParentID` = taxon_5.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_7 ON taxon_7.`ParentID` = taxon_6.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_8 ON taxon_8.`ParentID` = taxon_7.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_9 ON taxon_9.`ParentID` = taxon_8.`TaxonID` 
  LEFT OUTER JOIN taxon AS taxon_10 ON taxon_10.`ParentID` = taxon_9.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_1 ON determination_1.`IsCurrent` 
  AND determination_1.`CollectionMemberID` = 4 
  AND determination_1.`PreferredTaxonID` = taxon_1.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_2 ON determination_2.`IsCurrent` 
  AND determination_2.`CollectionMemberID` = 4 
  AND determination_2.`PreferredTaxonID` = taxon_2.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_3 ON determination_3.`IsCurrent` 
  AND determination_3.`CollectionMemberID` = 4 
  AND determination_3.`PreferredTaxonID` = taxon_3.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_4 ON determination_4.`IsCurrent` 
  AND determination_4.`CollectionMemberID` = 4 
  AND determination_4.`PreferredTaxonID` = taxon_4.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_5 ON determination_5.`IsCurrent` 
  AND determination_5.`CollectionMemberID` = 4 
  AND determination_5.`PreferredTaxonID` = taxon_5.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_6 ON determination_6.`IsCurrent` 
  AND determination_6.`CollectionMemberID` = 4 
  AND determination_6.`PreferredTaxonID` = taxon_6.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_7 ON determination_7.`IsCurrent` 
  AND determination_7.`CollectionMemberID` = 4 
  AND determination_7.`PreferredTaxonID` = taxon_7.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_8 ON determination_8.`IsCurrent` 
  AND determination_8.`CollectionMemberID` = 4 
  AND determination_8.`PreferredTaxonID` = taxon_8.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_9 ON determination_9.`IsCurrent` 
  AND determination_9.`CollectionMemberID` = 4 
  AND determination_9.`PreferredTaxonID` = taxon_9.`TaxonID` 
  LEFT OUTER JOIN determination AS determination_10 ON determination_10.`IsCurrent` 
  AND determination_10.`CollectionMemberID` = 4 
  AND determination_10.`PreferredTaxonID` = taxon_10.`TaxonID` 
WHERE 
  taxon_1.`ParentID` = 26 
  AND taxon_1.`TaxonTreeDefID` = 1 
GROUP BY 
  taxon_1.`TaxonID`
```
can be replaced by a recursive-cte like this:
```
WITH RECURSIVE anon_1(top_id, top_or_descendent_id, depth) AS (
  SELECT 
    taxon_1.`TaxonID` AS top_id, 
    taxon_1.`TaxonID` AS top_or_descendent_id,
    1 as depth
  FROM 
    taxon AS taxon_1 
  WHERE 
    taxon_1.`ParentID` = 26
    AND taxon_1.`TaxonTreeDefID` = 1
  UNION ALL 
  SELECT 
    anon_1.top_id AS top_id, 
    taxon_2.`TaxonID` AS top_or_descendent_id,
    depth + 1 as depth
  FROM 
    anon_1 
    INNER JOIN taxon AS taxon_2 ON taxon_2.`ParentID` = anon_1.top_or_descendent_id and anon_1.depth <= 9
) 
SELECT 
  anon_1.top_id AS anon_1_top_id, 
  CAST(
    sum(
      CASE WHEN (
        anon_1.top_or_descendent_id = anon_1.top_id 
        AND determination_1.`DeterminationID` IS NOT NULL
      ) THEN 1 ELSE 0 END
    ) AS SIGNED INTEGER
  ) AS anon_2, 
  count(
    determination_1.`DeterminationID`
  ) AS count_1 
FROM 
  anon_1 
  LEFT OUTER JOIN determination AS determination_1 ON determination_1.`IsCurrent` 
  AND determination_1.`PreferredTaxonID` = anon_1.top_or_descendent_id 
  AND determination_1.`CollectionMemberID` = 4
GROUP BY 
  anon_1.top_id
```

the depth clause is not entirely necessary, but helps safeguard against cycles. 

For testing, make sure that tree stats are exactly same in the branch, and in production. I'll recommend testing this for accuracy in smaller datasets (like coldFish), and then test for speed-up in bigger datasets (can try rjb-madrid on test panel). To really beat this up, set this property to zero in remote preferences 
image

### CODE REVIEW

**Comment by realVinayak** [comment]
For reference, this is currently done just for taxon / geography as I need to understand more specifics about other trees before I implement this solution. 

Also, some SQL versions don't support CTE so that is why exception handling is done.

**Comment by realVinayak** [comment]
Some more specifics on why it peforms better

Consider a taxon tree with just 3 species (A, B, C) and each of those species have 3 subspecies (so total 9 nodes). Depth is two, so if taxon (t1) is joined back to taxon aliased as t0 on (t0.parentid == t1.taxonid), then we get

```
t1 t0
A  subspA_1
A  subspA_2
A  subspA_3
B  subspB_1
B  subspB_2
B  subspB_3
C  subspC_1
C  subspC_2
C  subspC_3
```

Notice how t1 gets duplicated - each of which will be a separate join to determination (which it then takes distinct over). Flat-join makes determination join to both t1, and to t0

Using recursive-cte,
it first makes a subquery that results in
```
topid top_or_descendentid
A  A
A  subspA_1
A  subspA_2
A  subspA_3
B  B
B  subspB_1
B  subspB_2
B  subspB_3
C  C
C  subspC_1
C  subspC_2
C  subspC_3
```
Now, we make the join to determination with just top_or_descendentid which is guaranteed to have only distinct elements (removing redundant joins). It is possible to get the same above result by using flat joins (by doing it multiple times and having a union all), For example, the below query is an example that utilizes flat joins (and is still considerably faster than previous method, but slightly slower than recursive cte). Cte took 0.6 seconds, while below method took 1.6 seconds (not that big a difference, so I'm thinking of implementing below since it keeps joins intact)

```
select sum(count_1) as count_1, sum(if(top_id = top_or_d_id, count_1, 0)) as count_2, top_id from (select 
  taxon_1.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_1 
  left join determination on determination.preferredTaxonid = taxon_1.taxonid 
  and determination.isCurrent 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_1.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_1 
  left join determination on determination.preferredTaxonid = taxon_1.taxonid 
  and determination.isCurrent 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_2.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_2 
  left join determination on determination.preferredTaxonid = taxon_2.taxonid 
  and determination.isCurrent 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_3.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_3 
  left join determination on determination.preferredTaxonid = taxon_3.taxonid 
  and determination.isCurrent 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_4.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_4 
  left join determination on determination.preferredTaxonid = taxon_4.taxonid 
  and determination.isCurrent 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_5.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_5 
  left join determination on determination.preferredTaxonid = taxon_5.taxonid 
  and determination.isCurrent 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_6.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_6 
  left join determination on determination.preferredTaxonid = taxon_6.taxonid 
  and determination.isCurrent 
  join taxon as taxon_5 on taxon_6.parentid = taxon_5.taxonid 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_7.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_7 
  left join determination on determination.preferredTaxonid = taxon_7.taxonid 
  and determination.isCurrent 
  join taxon as taxon_6 on taxon_7.parentid = taxon_6.taxonid 
  join taxon as taxon_5 on taxon_6.parentid = taxon_5.taxonid 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_8.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_8 
  left join determination on determination.preferredTaxonid = taxon_8.taxonid 
  and determination.isCurrent 
  join taxon as taxon_7 on taxon_8.parentid = taxon_7.taxonid 
  join taxon as taxon_6 on taxon_7.parentid = taxon_6.taxonid 
  join taxon as taxon_5 on taxon_6.parentid = taxon_5.taxonid 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_9.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_9 
  left join determination on determination.preferredTaxonid = taxon_9.taxonid 
  and determination.isCurrent 
  join taxon as taxon_8 on taxon_9.parentid = taxon_8.taxonid 
  join taxon as taxon_7 on taxon_8.parentid = taxon_7.taxonid 
  join taxon as taxon_6 on taxon_7.parentid = taxon_6.taxonid 
  join taxon as taxon_5 on taxon_6.parentid = taxon_5.taxonid 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid 
union all 
select 
  taxon_10.taxonid as top_or_d_id, 
  taxon_1.taxonid as top_id, 
  count(determinationid) as count_1 
from 
  taxon as taxon_10 
  left join determination on determination.preferredTaxonid = taxon_10.taxonid 
  and determination.isCurrent 
  join taxon as taxon_9 on taxon_10.parentid = taxon_9.taxonid 
  join taxon as taxon_8 on taxon_9.parentid = taxon_8.taxonid 
  join taxon as taxon_7 on taxon_8.parentid = taxon_7.taxonid 
  join taxon as taxon_6 on taxon_7.parentid = taxon_6.taxonid 
  join taxon as taxon_5 on taxon_6.parentid = taxon_5.taxonid 
  join taxon as taxon_4 on taxon_5.parentid = taxon_4.taxonid 
  join taxon as taxon_3 on taxon_4.parentid = taxon_3.taxonid 
  join taxon as taxon_2 on taxon_3.parentid = taxon_2.taxonid 
  join taxon as taxon_1 on taxon_2.parentid = taxon_1.taxonid 
where 
  taxon_1.parentid = 26 
group by 
  taxon_1.taxonid) as temp group by top_id;

```

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/taxon-tree-manipulations-lock-database/1309/2

**Comment by grantfitzsimmons** [comment]
@specify/ux-testing @specify/dev-testing 

Can we make sure this is tested ASAP?

**Review by melton-jason — APPROVED** [review]
Not the most familiar with SQLAlchemy, but from what I can gather the changes here look solid.

**Comment by melton-jason** [comment]
If more in-depth developer testing is desired, the endpoint signature is defined as: 
`/api/specify_tree/(?P<tree>\w+)/(?P<treedef>\d+)/(?P<parentid>\w+)/stats/$`

where `tree` is the name of the tree model, `treedef` is the id of the {tree}TreeDef, and `parentid` the id of the node to fetch the stats for. 
 
For example:
`http://localhost/api/specify_tree/taxon/1/1/stats/`

The returned response may look something like:
```json
[
  [
    2,
    0,
    1
  ],
  [
    3,
    0,
    44975
  ]
]
```
Where each inner list corresponds to the stats for a child of the node given by `parentid`. 
In that list, the first integer is the id of the child, the second corresponds to the direct count of records referencing the child (like determinations referncing taxon), and the thrid number is the "indirect" count: or the sum of the total number of records referencing any descendant of the child. 

An openapi description/specification should be added (at some time) to the endpoint, but that is not a blocking factor of this Pull Request.

**Comment by realVinayak** [comment]
> and the thrid number is the "indirect" count: or the sum of the total number of records referencing any descendant of the child.

Almost! The third also includes the direct count + indirect count

File: specifyweb/specify/tree_views.py [hunk 1]
```diff
@@ -175,58 +178,104 @@ def tree_stats(request, treedef, tree, parentid):
...
-        query = session.query(getattr(child, child._id)) \
-            .filter(child.ParentID == parentid) \
-            .filter(getattr(child, treedef_col) == int(treedef)) \
-            .group_by(getattr(child, child._id))
-
-        query = make_joins(depth, query)
+            session.query(sql.func.count(distinct(tree_node.rankId))).filter(
+                tree_node.rankId >= highest_rank)[0]
+        query = None
+        try:
```
Comment (realVinayak) [inline specifyweb/specify/tree_views.py:240]: TEST: Check on MySQL 5.7 if exception is correctly handled

**Review by CarolineDenis — APPROVED** [review]

**Comment by pashiav** [comment]
Instructions for testers? @realVinayak

File: specifyweb/specify/tree_views.py (hunk 1 above)
Comment (realVinayak) [reply specifyweb/specify/tree_views.py:241]: Tested.

**Comment by realVinayak** [comment]
UX Testing Instructions

1. Go to App Resources -> Remote Preferences
2. Look for entries that start with `TreeEditor.Rank.Threshold.` (there should be 4). They specify the rank below which the tree stats need to run. Do not modify them yet.
3. Open two specify instances (one for this branch, and one for production) both on the same database (do this step on a small database)
4. Open the tree viewer for a particular tree, and go below the rank which was defined in the second step. If you are on taxon tree, go below the rank defined in `TreeEditor.Rank.Threshold.Taxon`. Now, "counts" should be visible next to the node name. 
5. Verify that the counts are same in both specify instances (this branch and production). Additionally, this branch should return numbers faster.
6. Now, go back to the Remote Preferences file, and set all the rank threshold to 0.
7. Go back to the respective tree. Check if this branch returns the count faster than production.

Do the above for all the trees, including paleo

**Review by carlosmbe — APPROVED** [review]
Based on the UX Instructions that @realVinayak provided. It seems good.  If more testing or an edge case comes to mind. Feel free to notify me @carlosmbe 

- [x] Open the tree viewer for a particular tree, and go below the rank which was defined in the second step. If you are on taxon tree, go below the rank defined in TreeEditor.Rank.Threshold.Taxon. Now, "counts" should be visible next to the node name.
- [ ] Verify that the counts are the same in both specify instances (this branch and production). Additionally, this branch should return numbers faster.
- [x] Now, go back to the Remote Preferences file, and set all the rank threshold to 0.
- [x] Go back to the respective tree. Check if this branch returns the count faster than production.

**Review by CarolineDenis — APPROVED** [review]

**Review by pashiav — COMMENTED** [review]
I tested on both a smaller db: `calvertmarinemuseum`, and a larger db: `KU_Fish_2023`.
Screenshots and recordings are of `KU_Fish_2023`

Verified both instances had the same counts:
pr-before
edge-before


Significantly quicker, looks great!

**Review by pashiav — APPROVED** [review]
I tested on both a smaller db: `calvertmarinemuseum`, and a larger db: `KU_Fish_2023`.
Screenshots and recordings are of `KU_Fish_2023`

Verified both instances had the same counts:
pr-before
edge-before


After setting all the rank threshold to 0, saving, logging out/back in:
**PR results:**

https://github.com/specify/specify7/assets/118556884/19177fd8-8dcd-40f5-859a-c3ea5658f4d1


**Production results:**

https://github.com/specify/specify7/assets/118556884/dec82720-25a8-4af8-9b62-49e3f1116171





Both tests on the smaller db and larger db worked. Results are significantly quicker! Looks great!

File: specifyweb/stored_queries/tests.py [hunk 2]
```diff
@@ -21,6 +27,87 @@ def test_stringid_roundtrip_en_masse(self) -> None:
...
+            selects = [sqlalchemy.select([sqlalchemy.literal(column).label(columns[idx][0]) for idx, column in enumerate(row)]) for row
+                       in result_set]
+            # union all instead of union because rows can be duplicated in the original query,
+            # but still need to preserve the duplication
+            unioned = sqlalchemy.union_all(*selects)
+            # Tests will fail when migrated to different background. TODO: Auto-detect dialects
+            final_query = str(unioned.compile(compile_kwargs={"literal_binds": True, }, dialect=mysql.dialect()))
+            return final_query, ()
+
+        cls.test_session_context = models.make_session_context(Session)
```
Comment (melton-jason) [inline specifyweb/stored_queries/tests.py:84]: (optional) 
Just a formatting nitpick here. 
Can this line be moved before the `run_django_query` function and the sqlalchemy listens_for decorator? 

This is setting an attribute on the `cls` object: it may be more organized to have this with similar lines (i.e., after the cls.engine and Session) to maintain a distinct separation of behaviors.

Comment (melton-jason) [reply specifyweb/stored_queries/tests.py:84]: Or is the 'before_cursor_execute' event listener bound in the session context and thus having it after the event listener is needed?

File: specifyweb/specify/test_trees.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
+from specifyweb.specify import models
+from specifyweb.specify.api_tests import ApiTests
+from specifyweb.specify.tree_stats import get_tree_stats
+from specifyweb.stored_queries.tests import SQLAlchemySetup
+
+
+def get_table(name: str):
```
Comment (melton-jason) [inline specifyweb/specify/test_trees.py:7]: (optional) 
This is personal preference, but I would move this to a more 'general' file. 
Perhaps to `tests.py` if you wish to keep it maintained in the "testing" environment. 
This function may be helpful when writing other (WorkBench) tests, and I would not expect such a function in a `test_trees.py` file. 

I was surprised to find I could not find a similar function in the "main" non-test files (there is the `get_table` and `get_table_strict` on the datamodel object, but seemingly no equivalent for Django). 
I know in most places we are either directly accessing the models from the model object (`models.Collectionobject`), or using something like the implementation here via getattr. 
We could perhaps use this in those places as well, and handle things such as AttributeErrors better (which are not being handled at all currently).

**Review by melton-jason — APPROVED** [review]
Unit tests look good! 
There were some questions regarding formatting I raised, but the logic looks sound. 

I'm glad the backend unit tests are getting some love!

File: specifyweb/stored_queries/tests.py [hunk 4]
```diff
@@ -21,6 +27,87 @@ def test_stringid_roundtrip_en_masse(self) -> None:
...
+    engine = None
+    test_session_context = None
+
+    @classmethod
+    def setUpClass(cls):
+        # Django creates a new database for testing. SQLAlchemy needs to connect to the test database
+        super().setUpClass()
+        cls.test_sa_url = settings.SA_TEST_DB_URL
+        _engine = sqlalchemy.create_engine(cls.test_sa_url, pool_recycle=settings.SA_POOL_RECYCLE,
+                                  connect_args={'cursorclass': SSCursor})
```
[the remark this answers is not in the document]
Comment (realVinayak) [reply specifyweb/stored_queries/tests.py:59]: > Is it feasible to assume we would need to configure or create further SQLAlchemy engines and connections?

Well, currently, every class does make it's own separate engine and connection (it is being used in setUpClass)

Can't think of a usage for this right now, so will remove the variable.

File: specifyweb/stored_queries/tests.py (hunk 2 above)
Comment (realVinayak) [reply specifyweb/stored_queries/tests.py:84]: Re-ordering it should have no affect.

File: specifyweb/specify/test_trees.py (hunk 3 above)
Comment (realVinayak) [reply specifyweb/specify/test_trees.py:7]: I was surprised too! AttributeErrors are actually fine to not handle. If you want to perform an action with a table (that's what a model just a representation of) but that table doesn't exist, you can't do anything. So, just propagate that error up is a good choice.

**Review by melton-jason — APPROVED** [review]

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/specify-7-9-2-release-announcement/1452/1
